### PR TITLE
Make sure climate statistics always returns original dtype

### LINKED ISSUE #1235: The climate_statistics preprocessor sometimes changes the dtype
**Describe the bug**
It has been reported that the multi-model statistics can have a problem with inconsistent dtypes (see https://github.com/ESMValGroup/ESMValTool/discussions/2218#discussioncomment-992288 and https://github.com/ESMValGroup/ESMValTool/issues/2198#issuecomment-863908898).

This seems, however, to originate a bit higher up in the processing chain. In this case, the climate_statistics preprocessor sometimes changes the dtype of the incoming cube, leading to inconsistent dtypes in the input to the multi-model statistics.

Probably the best way to address this is to fix the climate_statistics preprocessor to produce output with a consistent dtype.

### PR DESCRIPTION
**Pull request by zklaus**
## Description

This adds code to the climate_statistics preprocessor to detect whether it has changed the dtype of the cube. If this is the case, it changes the dtype back to the original dtype.

Closes #1235 

Link to documentation:

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

File: esmvalcore/preprocessor/_time.py [hunk 1]
```diff
@@ -552,25 +553,32 @@ def climate_statistics(cube,
...
+        operator = get_iris_analysis_operation(operator)
+        clim_cube = cube.aggregated_by(clim_coord, operator)
+        clim_cube.remove_coord('time')
+        if clim_cube.coord(clim_coord.name()).is_monotonic():
+            iris.util.promote_aux_coord_to_dim_coord(clim_cube,
+                                                     clim_coord.name())
+        else:
+            clim_cube = iris.cube.CubeList(
+                clim_cube.slices_over(clim_coord.name())).merge_cube()
+        cube.remove_coord(clim_coord)
```
Comment (zklaus) [inline esmvalcore/preprocessor/_time.py:574]: This line looks a bit fishy to me, but it was there before, so there is that. It looks like this should be `clim_cube.remove_coord(clim_coord)`. @valeriupredoi, this seems to come from you. Do you have an opinion?

**Comment by jvegasbsc** [comment]
I am changing the tests in the time module to always check for the types

**Comment by jvegasbsc** [comment]
I have detected the same problem when computing standardized anomalies, as they get converted to `float64` from a `float32` input.  I am trying to solve it and then I will push the changes

**Comment by zklaus** [comment]
> I have detected the same problem when computing standardized anomalies, as they get converted to `float64` from a `float32` input. I am trying to solve it and then I will push the changes

Be advised that I came to this solution via the anomalies. Turns out the problem with anomalies was that the climate_statistics, which are used in the calculation of the anomalies did this conversion, so maybe once this PR is in, the anomalies problem vanishes.

**Comment by jvegasbsc** [comment]
> Be advised that I came to this solution via the anomalies. Turns out the problem with anomalies was that the climate_statistics, which are used in the calculation of the anomalies did this conversion, so maybe once this PR is in, the anomalies problem vanishes.

I am working directly on this PR, so that's not the issue

**Review by jvegasbsc — APPROVED** [review]
Nice catch! I currently do not have access to data (BSC annual maintenace)  to properly test with a recipe, though.

**Pull request merged**